### PR TITLE
Rhel8 alpha branch add nvdimm support

### PR DESCRIPTION
**Pull request by rvykydal**
Related PRs:
https://github.com/storaged-project/blivet/pull/658
https://github.com/clumens/pykickstart/pull/214/ -> https://github.com/clumens/pykickstart/pull/215

### CODE REVIEW

**Comment by rvykydal** [comment]
Screencast of GUI changes:

- only sector mode devices can be selected for installation

- reconfiguration to sector mode can be done in NVDIMM tab

- selection of devices for reconfiguration is done in the tab treeview

- TODO: after reconfiguring the devcietree needs to be updated but we can't do just populate here to add the new sector mode device, we have to do also reset to remove the original memory mode device from the tree. I am still not sure what to call here (deivicetree.reset and devicetree.populate or storage.reset, or even storage_initialize?). Anyway the repopulating takes quite a long time so I plan to run it in thread and fire up a spinner and a message about the storage reset in the dialog.

- At the end of the video there is filtering by namespace which should be changed to filter by prefix to be more valuable I think

- https://rvykydal.fedorapeople.org/nvdimm/nvdimm.webm

**Comment by rvykydal** [comment]
TODO:
need to add data loss warning to reconfiguration in GUI

File: pyanaconda/ui/gui/spokes/advanced_storage.py [hunk 1]
```diff
@@ -774,6 +789,20 @@ def on_add_dasd_clicked(self, widget, *args):
         # storage are displayed in the UI.
         self.refresh()
 
+    def on_reconfigure_nvdimm_clicked(self, widget, *args):
+
+        namespaces = self.pages[3].get_selected_namespaces()
+
+        dialog = NVDIMMDialog(self.data, self.storage, namespaces)
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/advanced_storage.py]: Is here a reason to have those two blank lines above? I think you can remove them.

File: pyanaconda/ui/gui/spokes/advstorage/nvdimm.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        # marked insensitive.
+        self._waitSpinner.stop()
+
+        if self._error:
+            self.builder.get_object("deviceErrorLabel").set_text(self._error)
+            self._error = None
+            self._okButton.set_sensitive(True)
+            self._conditionNotebook.set_current_page(PAGE_RESULT_ERROR)
+        else:
+            self._okButton.set_sensitive(True)
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/advstorage/nvdimm.py]: I think you can merge these two calls of `self._okButton.set_sensitive(True)` outside of that if-else branch.  Of course if that shouldn't be `False` in one branch.

File: pyanaconda/storage/osinstall.py [hunk 3]
```diff
@@ -37,7 +37,7 @@
 from blivet import arch, util, udev
 from blivet.blivet import Blivet
 from blivet.storage_log import log_exception_info
-from blivet.devices import FileDevice, NFSDevice, NoDevice, OpticalDevice, NetworkStorageDevice, DirectoryDevice, MDRaidArrayDevice, PartitionDevice, BTRFSSubVolumeDevice, TmpFSDevice, LVMLogicalVolumeDevice, LVMVolumeGroupDevice, BTRFSDevice
+from blivet.devices import FileDevice, NFSDevice, NoDevice, OpticalDevice, NetworkStorageDevice, DirectoryDevice, MDRaidArrayDevice, PartitionDevice, BTRFSSubVolumeDevice, TmpFSDevice, LVMLogicalVolumeDevice, LVMVolumeGroupDevice, BTRFSDevice, NVDIMMNamespaceDevice
```
Comment (jkonecny12) [inline pyanaconda/storage/osinstall.py]: Please wrap this line to 99 characters.

File: pyanaconda/ui/gui/spokes/advanced_storage.py [hunk 4]
```diff
@@ -498,8 +498,13 @@ def setup(self, store, selectedNames, disks):
         for disk in disks:
             paths = [d.name for d in disk.parents]
             selected = disk.name in selectedNames
+            mutable = not disk.protected
 
-            store.append([True, selected, not disk.protected,
+            if disk.mode != "sector":
```
Comment (jkonecny12) [inline pyanaconda/ui/gui/spokes/advanced_storage.py]: Are you sure that the `mode` attribute will be here all the time? If not please add the `hasattr(disk, 'mode')` check.

**Comment by rvykydal** [comment]
TODO:
- ignore existing sector mode devices unless explicitly required by nvdimm command

**Comment by rvykydal** [comment]
Applied remarks from @jkonecny12 .

**Comment by rvykydal** [comment]
TODO:
nvdimm command in kickstart generated by anaconda

**Comment by rvykydal** [comment]
I've added one more patch for "nvdimm use" command.

**Comment by rvykydal** [comment]
Just added some missing contexts for translators to glade files (CI).

**Comment by rvykydal** [comment]
jenkins, test this please

**Comment by rvykydal** [comment]
jenkins, test this please

File: pyanaconda/ui/gui/spokes/advanced_storage.py [hunk 5]
```diff
@@ -473,13 +475,88 @@ def visible_func(self, model, itr, *args):
...
+    def visible_func(self, model, itr, *args):
+        obj = DiskStoreRow(*model[itr])
+        device = self.storage.devicetree.get_device_by_name(obj.name, hidden=True)
+        return self.ismember(device) and self._filter_func(device)
+
+    def get_selected_namespaces(self):
+        selection = self.treeview.get_selection()
+        store, pathlist = selection.get_selected_rows()
+        namespaces = [store[store.get_iter(path)][17]
+                      for path in pathlist]
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/gui/spokes/advanced_storage.py]: Done

File: pyanaconda/ui/gui/spokes/advanced_storage.py [hunk 6]
```diff
@@ -700,6 +789,18 @@ def on_add_dasd_clicked(self, widget, *args):
         # storage are displayed in the UI.
         self.refresh()
 
+    def on_reconfigure_nvdimm_clicked(self, widget, *args):
+        namespaces = self.pages[3].get_selected_namespaces()
```
[the remark this answers is not in the document]
Comment (rvykydal) [reply pyanaconda/ui/gui/spokes/advanced_storage.py]: Done

**Comment by rvykydal** [comment]
Added 3 patches.

- 1. and 2.: port of bug https://bugzilla.redhat.com/show_bug.cgi?id=1562301 of which the 2. improves bootloader error messaging in UI

- 3: disallowing boot on nvdimm, the patch is using the improved error messaging

File: pyanaconda/bootloader.py [hunk 7]
```diff
@@ -674,9 +678,15 @@ def is_valid_stage2_device(self, device, linux=True, non_linux=False):
...
-        if _is_on_sw_iscsi(device) and not _is_on_ibft(device):
-            self.errors.append(_("%s cannot be on an iSCSI disk") % self.stage2_description)
-            valid = False
+        if _is_on_sw_iscsi(device):
+            if not _is_on_ibft(device):
+                if flags.nonibftiscsiboot:
+                    log.info("%s on non-iBFT iSCSI disk allowed by boot option inst.nonibftiscsiboot",
+                             self.stage2_description)
+                else:
+                    self.errors.append(_("%(bootloader_stage2_description)s cannot be on an iSCSI disk which is not configured in iBFT.")
```
Comment (jkonecny12) [inline pyanaconda/bootloader.py]: Please wrap this line to 99 character.

File: pyanaconda/bootloader.py [hunk 8]
```diff
@@ -568,6 +568,7 @@ def is_valid_stage1_device(self, device, early=False):
                     log.debug("stage1 device on non-iBFT iSCSI disk allowed by boot option inst.iscsi.nonibftboot")
                 else:
                     log.debug("stage1 device cannot be on an non-iBFT iSCSI disk")
+                    self.errors.append(_("Boot loader stage1 device cannot be on an iSCSI disk which is not configured in iBFT."))
```
Comment (jkonecny12) [inline pyanaconda/bootloader.py]: Also this line.

File: pyanaconda/bootloader.py (hunk 7 above)
Comment (rvykydal) [reply pyanaconda/bootloader.py]: Done

File: pyanaconda/bootloader.py (hunk 8 above)
Comment (rvykydal) [reply pyanaconda/bootloader.py]: Done

**Comment by jkonecny12** [comment]
Looks good to me now.

**Comment by rvykydal** [comment]
jenkins, test this please

**Comment by rvykydal** [comment]
Added three more patches:
One is fixing ignoring of nvdimm devices.
The other two are for UI - adding data loss warning and spinner for storage reset. For those two see the screencasts:
current state: https://rvykydal.fedorapeople.org/nvdimm/nvdimm_reconfigure.webm
patched: https://rvykydal.fedorapeople.org/nvdimm/nvdimm_reconfigure_with_spinner_and_warning.webm

**Comment by rvykydal** [comment]
Just bumping required pykickstart and python-blivet versions

**Comment by rvykydal** [comment]
Ported to master with https://github.com/rhinstaller/anaconda/pull/1472

**Pull request merged**